### ISSUE TITLE
The decision head's output is unnormalized (~300x the encoder's scale), saturating the act head so act_probability is always 1.0

### ISSUE THREAD

**Issue by ashaazami** [NandhaKishorM/laya#185]
Problem checked against `laya` 0.3.5 (`573e5b6`) and the checkpoints on Hugging Face on 2026-09-22.

`DecisionModel`'s head layers are pre-norm (`norm_first=True`) and nothing normalizes the residual
stream afterwards, so activations leaving the head are about 300× the scale of the encoder's
output. `scorer` is unaffected because it starts with its own `nn.LayerNorm`, so the option
probabilities are fine. `pooled` (the `[CLS]` vector) goes straight into `act_head`, which has no
normalization, so it saturates and `action.act_probability` is 1.0 for every input.

```python
self.scorer   = nn.Sequential(nn.LayerNorm(d), nn.Linear(d, d), nn.GELU(), nn.Linear(d, 1))
self.act_head = nn.Sequential(nn.Linear(d + 4, 256), nn.GELU(), nn.Linear(256, n_act))
```

## Scale through the model

One request (`convaiinnovations/laya`, a 3-option choice question about "I was charged twice,
refund please."):

```
encoder output   : |CLS| 35.6      max|h| 26.1      mean|h| 0.47
after head layers: |CLS| 10838.0   max|h| 1238.6    mean|h| 220.44
act logits       : [4499.5, -3677.2]      -> softmax [1.0, 0.0]
```

The act head's own weights are small (max |w| = 0.08); it is its input that is out of range.

Measured with:

```python
import torch, laya
from laya.common import build_sequence, collate_items, QTYPES

agent = laya.load("convaiinnovations/laya")
q = {"t": "choice", "ins": "Topic?", "crit": {"billing": None, "tech": None, "other": None}}
ids, markers = build_sequence(agent.tok, {"msg": "I was charged twice, refund please."}, q, 512, 192)
b = collate_items([[{"ids": ids, "markers": markers, "qtype": QTYPES["choice"]}]], agent.tok.pad_token_id)
m = agent.model
with torch.no_grad():
    h = m.encoder(input_ids=b["input_ids"].to(agent.device),
                  attention_mask=b["attention_mask"].to(agent.device)).last_hidden_state
    print("encoder output   :", h[0, 0].norm().item(), h.abs().max().item(), h.abs().mean().item())
    h2 = h + m.type_emb(b["qtype"].to(agent.device))[:, None, :]
    for layer in m.head.layers:
        h2 = layer(h2, src_key_padding_mask=~b["attention_mask"].bool().to(agent.device))
    print("after head layers:", h2[0, 0].norm().item(), h2.abs().max().item(), h2.abs().mean().item())
```

## What a caller sees

`act_probability` is exactly 1.0 for every input I have tried: English and multilingual states, all
three question types, an empty state, ambiguous options, and nonsense text. `confidence` moves as
expected across the same inputs, which makes the contrast clear:

```python
import laya
agent = laya.load("convaiinnovations/laya")
questions = {"q": {"type": "choice", "instructions": "Topic?", "criteria": ["billing", "tech", "other"]}}
for state in ["I was charged twice, refund please.", "", "qwx zzz 1234"]:
    a = agent.predict(state, questions)["answers"]["q"]
    print(a["confidence"], a["action"]["act_probability"])
```

| state | confidence | `act_probability` |
|---|---:|---:|
| "I was charged twice, refund please." | 0.7308 | 1.0 |
| "" (empty) | 0.0082 | 1.0 |
| "qwx zzz 1234" | 0.0437 | 1.0 |

## The head still responds to the input

The raw act logits do vary with the input; the offset between the two classes simply dwarfs that
variation, so the softmax cannot express it:

| state | raw act logits `[act, escalate]` | gap |
|---|---|---:|
| "I was charged twice, refund please." | [4701.5, -3841.8] | 8543.3 |
| "" (empty) | [4509.8, -3678.5] | 8188.3 |
| "qwx zzz 1234" | [4446.7, -3630.1] | 8076.8 |

The gap moves in the expected direction (widest for the clear case, narrowest for nonsense) across
a range of ~470, against an offset of ~8,000. The signal exists; the scale hides it.

## Questions

- Is `act_probability` meant to be usable, or is it a leftover from the RL objective that callers
  should ignore? If the latter, documenting that would save readers the investigation.
- If it is meant to be usable: a `LayerNorm` in front of `act_head`, or a final norm after the head
  layers, would bring its input back into range. Note that the option logits are also read from the
  same unnormalized stream (through `scorer`'s own `LayerNorm`), so a norm added after the head
  layers changes the published checkpoints' numerics and would need refitting. A temperature on the
  act logits alone (they currently get none, unlike the option logits) would recover a usable range
  without touching the weights.

**Comment by Ashfaqbs** [comment]
Happy to take a pass at this if it's useful, but the two directions have different blast radii and I'd rather confirm which one you want before opening a PR:

- **Fixed temperature on `act_logits` only** (no `LayerNorm`, no touching `act_head`'s weights): a single scalar divides `act_logits` before the softmax, tuned so `act_probability` spans a usable range on the measurements above. Doesn't change `option`/`score`/`noul` numerics at all (those already go through `scorer`'s own `LayerNorm`), and doesn't require the published checkpoints to be refit — it's a post-hoc calibration, not an architecture change.
- **`LayerNorm` before `act_head`** (or after the head layers): fixes the saturation at the source, but changes the numerics `act_head` was trained against, so the shipped checkpoints would read differently through it until refit — not something I'd want to ship without you confirming that tradeoff is acceptable.

Given the gap is ~8000 vs a signal range of ~470, a fixed temperature alone gets `act_probability` back into a range where it's informative rather than a constant 1.0, without any risk to the option/score/noul outputs or the existing checkpoints. If that's the direction you'd want, I can put together the PR (temperature as a `DecisionModel` constructor arg with a measured default, calibrated against a spread of inputs rather than the one example above) plus a regression test asserting `act_probability` actually varies across a set of clear/ambiguous/empty states.

If `act_probability` is closer to a leftover from the RL objective that callers shouldn't rely on, happy to just help get that documented instead so the next person doesn't lose time on it.

**Comment by NandhaKishorM** [comment]
Thank you @Ashfaqbs for laying out both options so clearly. I measured them on the real English checkpoint before answering, using 396 labelled decisions.

- **LayerNorm before the head (#192):** the logit gap drops from about 8,000 to about 24, but `act_probability` is still 1.0 for every input, including an empty state and keyboard noise.
- **A temperature on the raw logits:** the raw gap does vary (2,679 to 8,857), but it runs *against* correctness, with an AUROC of 0.30. The head leans toward "act" more strongly when the answer is wrong. A temperature would make `act_probability` move, but in the wrong direction. `confidence` gets 0.77 on the same items.

So your last option is the right one. `act_probability` is a leftover the current checkpoints never learned, and it should be documented rather than calibrated. That note shipped in the README's Limits section in 0.3.7 today. The real fix is retraining the act head. Keeping this open to track that. Thank you for offering to take it on.

**Comment by Ashfaqbs** [comment]
Thanks for measuring both options before deciding. The AUROC of 0.30 for the raw gap is what settles it for me: a temperature would make `act_probability` move, but in the wrong direction, so documenting it is the honest call. I'll leave the retraining of the act head to you. If you ever want a second pair of hands on an eval harness for it (for example scoring `act_probability` against `confidence` on labelled decisions), I'm happy to help.

**Comment by NandhaKishorM** [comment]
Thank you @Ashfaqbs. I've closed #192, because normalising at inference changes the released checkpoints' act outputs without making them informative, as the measurement showed. This issue stays open to track the act head. An eval harness scoring `act_probability` against labelled decisions would be a welcome research contribution, since it gives any future fix a clear bar to beat.
